web/components/portfolio/portfolio-chart.tsx

```diff
@@ -0,0 +1,282 @@
+import { HistoryPoint, ValueKind } from 'common/chart'
+import { formatMoneyNumber } from 'common/util/format'
+import { axisBottom, axisRight } from 'd3-axis'
+import { ScaleContinuousNumeric, ScaleTime } from 'd3-scale'
+import { CurveFactory, curveStepAfter } from 'd3-shape'
+import { mapValues } from 'lodash'
+import { ReactNode, useCallback, useEffect, useMemo, useState } from 'react'
+import { useEvent } from 'web/hooks/use-event'
+import { dataAtTimeSelector, dataAtXSelector } from '../charts/generic-charts'
+import {
+  AreaPath,
+  LinePath,
+  PointerMode,
+  SVGChart,
+  SliceMarker,
+  TooltipProps,
+  ZoomParams,
+} from '../charts/helpers'
+import { ZoomSlider } from '../charts/zoom-slider'
+import clsx from 'clsx'
+import { useEffectCheckEquality } from 'web/hooks/use-effect-check-equality'
+import { BALANCE_COLOR, PortfolioMode } from './portfolio-value-graph'
+import { PortfolioHoveredGraphType } from './portfolio-value-section'
+import { Col } from '../layout/col'
+import dayjs from 'dayjs'
+import { CoinNumber } from '../widgets/manaCoinNumber'
+
+type AreaPointType = {
+  x: number // The x-coordinate
+  y0: number // Lower boundary of the area
+  y1: number // Upper boundary of the area
+}
+
+// hacky solution
+const BALANCE_IDX = 0
+const INVESTED_IDX = 1
+
+// multi line chart
+export const PortfolioChart = <P extends HistoryPoint>(props: {
```

Comment:
This looks good. It's a shame there's not a good way to build a simpler generic version of this stacked line graph, and then configure it for the portfolio case? It would certainly be cleaner to read if so.